pkg/endpoint/policy.go

```diff
@@ -429,24 +430,38 @@ func (e *Endpoint) Regenerate(owner Owner, regenMetadata *ExternalRegenerationMe
 			e.getLogger().Debug("Dequeued endpoint from build queue")
 
 			regenContext.DoneFunc = doneFunc
-			err := e.regenerate(owner, regenContext)
-			doneFunc() // in case not called already
 
-			repr, reprerr := monitorAPI.EndpointRegenRepr(e, err)
-			if reprerr != nil {
-				e.getLogger().WithError(reprerr).Warn("Notifying monitor about endpoint regeneration failed")
-			}
-
-			if err != nil {
-				buildSuccess = false
-				if reprerr == nil && !option.Config.DryMode {
-					owner.SendNotification(monitorAPI.AgentNotifyEndpointRegenerateFail, repr)
-				}
-			} else {
-				buildSuccess = true
-				if reprerr == nil && !option.Config.DryMode {
-					owner.SendNotification(monitorAPI.AgentNotifyEndpointRegenerateSuccess, repr)
+			epEvent := eventqueue.NewEvent(&EndpointRegenerationEvent{
```

Comment:
I think this could be a lot simple with the usage of `pkg/serializer`

```golang
			s := serializer.FunctionQueue{}
			s.Enqueue(func() error {
				err = e.regenerate(owner, regenContext)
				var regenError error
				owner.QueueEndpointBuild(uint64(e.ID))
				// Notify monitor of result of regeneration.
				e.notifyEndpointRegeneration(owner, regenError)
				return nil
			}, serializer.NoRetry)
```

Reply:
`pkg/serializer` would have to be edited to contain the following:

* Cancellation of events in the queue so the caller can be aware that they were cancelled
* Logic that EventQueue has to not Enqueue events when the channel is closed 
* Logic to handle potential for multiple closes
* Generic result handling (for future work if needed)
* Draining of events from the queue

I think that the flexibility that EventQueue offers could even replace `pkg/serializer` over time.

Reply:
> `pkg/serializer` would have to be edited to contain the following:
> 
> * Cancellation of events in the queue so the caller can be aware that they were cancelled.

That can already be done with context

```golang
 			ctx := context.WithTimetout(context.WithBackground(), 10*time.Second)
			s := serializer.FunctionQueue{}
			s.Enqueue(func() error {
				select {
					case <- ctx.Done():
					return errors.New("context deadline")
					default:
				}
				err = e.regenerate(owner, regenContext)
				var regenError error
				owner.QueueEndpointBuild(uint64(e.ID))
				// Notify monitor of result of regeneration.
				e.notifyEndpointRegeneration(owner, regenError)
				return nil
			}, serializer.NoRetry)
```

or even passing a context argument to a new function called `serializer.EnqueueWithContext`

> * Logic that `EventQueue` has to not Enqueue events when the channel is closed.

Already exists with `serializer.FunctionQueue{}.Stop()`

> * Logic to handle potential for multiple closes of the `EventQueue`.

Why would you want to perform multiple closes? That can be added with a sync.Do inside the pkg/serializer

> * Generic result handling instead of just returning an error.

The error returned is for the `pkg/serializer` to retry again in case of failure. Why can't the result handling of the event be immediately done after executing the gist of the function?

```golang
			s.Enqueue(func() error {
				result := myMainFunc()
				resultHandler(result)
			}, serializer.NoRetry)
```


> * Draining of events from the queue.

`serializer.FunctionQueue{}.Stop()`?

> 
> I think that the flexibility that EventQueue offers could even replace `pkg/serializer` over time, or its functionality could be put into `pkg/serializer`. I do agree that the handling of the `cancelled` or `result` cases could be a bit cleaner / simpler, like what I have described here:
> https://github.com/cilium/cilium/pull/6943/files/d816971e18db5ca08877cd70735022431c68edd1#r258734360

Reply:
@aanm 

> That can already be done with context.

I don't see how your proposal with the usage of context handles long-running events. I want to be able to clearly signal to the caller that the event was cancelled.  

> Already exists with `serializer.FunctionQueue{}.Stop()`

If we call that function twice, it will panic - this can be changed.

> The error returned is for the `pkg/serializer` to retry again in case of failure. Why can't the result handling of the event be immediately done after executing the gist of the function?

I designed the queue so that it only processes the given event which needs to be blocking, e.g., regeneration of an endpoint, and that the handling of the result of that event should be done by the queuer of said event, as to not block the queue from processing more events.  The critical section of the queue should be as small as possible, in my opinion. Result handling in the future could potentially be costly as well in terms of acquiring locks of locks, waiting, etc. 

> serializer.FunctionQueue{}.Stop()

(this was in reference to draining the queue) - there is no way to signal to the events here whether they were processed or not - my rationale for having results processed by the queuer is explained above.  The 'cancelling' of queued events is nice for the caller to get visibility into what happens. This is a big differentiator between this and `pkg/serializer`, which queues functions up without waiting for a specific result.

Reply:
> @aanm
> 
> > That can already be done with context.
> 
> I don't see how your proposal with the usage of context handles long-running events. I want to be able to clearly signal to the caller that the event was cancelled.
> 

I don't understand, can you give an example of who will be the caller of what?

> > Already exists with `serializer.FunctionQueue{}.Stop()`
> 
> If we call that function twice, it will panic - this can be changed.
> 
> > The error returned is for the `pkg/serializer` to retry again in case of failure. Why can't the result handling of the event be immediately done after executing the gist of the function?
> 
> I designed the queue so that it only processes the given event which needs to be blocking, e.g., regeneration of an endpoint, and that the handling of the result of that event should be done by the queuer of said event, as to not block the queue from processing more events. The critical section of the queue should be as small as possible, in my opinion. Result handling in the future could potentially be costly as well in terms of acquiring locks of locks, waiting, etc.
> 

If the queue only processes the given event why do we have a queue in the first place? Why isn't simply a normal code flow without any queue involved? Let me ask you this, what is the difference between the 2 codes:

```golang
func dostuff(){
        result := run()
        handleResult(result)
}
```

and with queue.

```golang
func dostuff(){
        eh := &evenhandler{}
        queueStuff(eh)
        result := <-eh.Result  //blocking operation
        handleResult(result)
}
```

to me there is none.

> > serializer.FunctionQueue{}.Stop()
> 
> (this was in reference to draining the queue) - there is no way to signal to the events here whether they were processed or not - my rationale for having results processed by the queuer is explained above. The 'cancelling' of queued events is nice for the caller to get visibility into what happens. This is a big differentiator between this and `pkg/serializer`, which queues functions up without waiting for a specific result.

This can be achieved with the golang context, there's no need to reinvent the wheel.

```golang
// returns a new context where the caller will know if the context was closed
ctx := queuer.EnqueueWithContext(func (ctx context.Context) {
		select {
		case <-ctx.Done():
			return // returning has context was canceled
		default:
		}
		// do other stuff
})

queuer.Stop() // will drain the queue and cancel all contexts

select {
case <-ctx.Done():
	fmt.Println("My request was canceled")
default:
	fmt.Println("My request was not canceled")
}
```

Reply:
> I don't understand, can you give an example of who will be the caller of what?

The queuer queues up an event that must be done in a mutually exclusive manner in relation to other events on the EventQueue. The EventQueue performs that event, and sends the result of that event back to the queuer so the queuer can handle the result of the event in a means that they desire. 

> If the queue only processes the given event why do we have a queue in the first place? Why isn't simply a normal code flow without any queue involved? 

The queue blocks *other* queued events from running concurrently as the given event which is being processed by the queue. This is undesirable, because we need to make sure that we are not bumping the policy revision prematurely while we are regenerating an endpoint; these events must be mutually exclusive, which is what the `EventQueue` ensures. 

> This can be achieved with the golang context, there's no need to reinvent the wheel.

Would I still be able to send results back to the queuer with the use of a context? I would like to maintain that flexibility. That was the primary reason why I designed the `EventQueue` in the fashion that I did. I think having the flexibility of being able to send results back to the caller is nice, as it ensures that the critical section of events being processed in the `EventQueue` is kept to a minimum.

Reply:
@aanm I understand your argument and I typically agree that we should not introduce frameworks unless required. That said, a couple of points why I prefer Ian's framework and why I think it makes sense to invest:

1. The serializer is a minimal version of the EventQueue, we would evolve the serializer into the EventQueue anyway. A good example here is the introduction of metrics to measure the latency between Enqueue and execution with break downs for time spent waiting and time spent executing. Passing in a pure function will not allow producing good metrics so replacement of the function with a structure or interface as context will happen anyway.
2. While the first implementation of EventQueue is blocking, this will evolve into an async event queue soon. We should not disregard the framework of the event queue simply because the first revision does not fully utilize all potential yet. This event queue will be absolutely crucial for performance and scale and having a lot of visibility, fairness control and resource usage control will become absolutely essential.
3. We should not encourage more open coding of things like context and timeout. The open coding of context in endpoint creation via the API is hurting us a lot due to generic errors and lack of visibility with metrics.

My proposal is to evolve the event queue with the use case of the endpoint in mind and fold in the serializer later on.

Reply:
@tgraf then why don't we start with the `serializer` and evolve it? The EventQueue is lacking metric visibility as well as the `serializer`, adding metrics into the serializer would be really simple and we would immediately gain visibility of `CiliumNetworkPolicy`, `Services`, `NetworkPolicy`, `Endpoint` and `Pods` waiting time and execution of handling those events. Instead of forking 2 frameworks that accomplish the exact same behavior why not don't we modify the existing one, which does not require that many modifications, to accommodate the new use cases?